CRM-21037 activity sendsms unittests

mattwire opened the pull request:
Overview
----------------------------------------
Add unit tests for Send SMS functionality in core.

Before
----------------------------------------
No test coverage.

After
----------------------------------------
Test coverage for Activity BAO functions.

Technical Notes
----------------------------------------
This adds special code to handle unit tests in CRM/Sms/Provider.php for a provider with name "testSmsProvider".  The core code expects ALL SMS providers to be extensions, the unit test one is not and we would get a fatal error if we didn't modify core here.

Comments
----------------------------------------
Some variables renamed, function cleanup, more CRM_Core_Pseudoconstant.  Unit test should be run before applying the rest of the changes and again afterwards to demonstrate there are no changes.

---

 * [CRM-21037: Add unit tests for Activity sendSMS functions](https://issues.civicrm.org/jira/browse/CRM-21037)

mattwire commented:
Jenkins test this please

totten commented on tests/phpunit/CiviTest/CiviUnitTestCase.php:
```diff
@@ -3915,3 +3915,42 @@ public function getFormObject($class) {
+/**
+ * Test SMS provider to allow for testing
+ */
+class testSMSProvider extends CRM_SMS_Provider {
```
Zoink? `CiviUnitTestCase.php` is s strange place to put this?

mattwire replied:
@totten Yes it is a strange place - any suggestions please?  It was in the tests/phpunit/CRM/Sms/Provider.php file.  I didn't really want to duplicate the class so needed to put it somewhere that it could be discovered by both sets of tests. CiviUnitTestCase.php seems to be the only common denominator!

mlutfy replied:
@mattwire Would something similar to `CiviTest/CiviMailUtils.php` work out?

michaelmcandrew replied:
@mattwire @totten @mlutfy - I pushed a commit which reverts the change to this file `and` instead creates a new `tests/phpunit/CiviTest/CiviTestSMSProvider.php` file. Annoyingly(?), it seems to squash this thread / presume it is no longer needed as part of the discussion.

michaelmcandrew commented:
@mattwire @totten @mlutfy. You participated in a thread which got hidden because a commit I made to this branch reverted the changes to the file where the discussion was attached. For reference, you can view the thread [here](https://github.com/civicrm/civicrm-core/pull/10946/commits/85074206d75f4079d300be65f22d189c18b1c17c#r137335653).

Now that you have got your heads around that :), following @mlutfy's train of thought, I moved and renamed the class for the test provider from CiviUnitTestCase.php to CiviTestSMSProvider.php and explicitly required it when necessary.

Hoping that makes this more mergable and we can get some more SMS tests into core.

seamuslee001 commented:
@michaelmcandrew can you rebase this please?

mattwire commented:
@seamuslee001 Squashed and rebased. @michaelmcandrew I've squashed your changes into mine, should be easier to review now, but that does lose your commit message :-(

michaelmcandrew commented:
@mattwire - not a problem :)

seamuslee001 commented on tests/phpunit/CRM/Activity/BAO/ActivityTest.php:
```diff
@@ -1135,4 +1135,140 @@ public function testSendEmailWithCampaign() {
+      'is_active'  => 1,
+      'version'    => 3,
+    );
+    $providerValue = civicrm_api3('option_value', 'create', $params);
```
@mattwire This option value isn't properly being torn down me thinks following the test causing the test failures

seamuslee001 commented:
ok I have reviewed the code and it looks good to me, there is an issue with the test tho not properly being cleaned up causing subsequent test failures. The code changes are mailing improving the naming of various Variables and using the correct functions to get option value ids,

seamuslee001 commented on CRM/Activity/BAO/Activity.php:
```diff
@@ -1638,73 +1638,81 @@ public static function sendEmail(
   ) {
-    if ($userID == NULL) {
-      $userID = CRM_Core_Session::getLoggedInContactID();
+    if (!isset($contactDetails) && !isset($contactIds)) {
```
@mattwire do you think we should be doing `empty()` instead, just thinking what if both params are supplied but they are both empty are we ok with that?

seamuslee001 replied:
@mattwire i feel like maybe now we should put this below the permissions check given that probably should be the first thing we check

seamuslee001 commented on CRM/Activity/BAO/Activity.php:
```diff
@@ -1638,73 +1638,81 @@ public static function sendEmail(
+      Throw new CRM_Core_Exception('You must specify either $contactDetails or $contactIds');
+    }
+    // Populate $contactDetails and $contactIds if only one is set
+    if (is_array($contactIds) && !empty($contactIds) && empty($contactDetails)) {
```
if `$contactIds` is not an array what will it be just a single id or will it be a `,` separated string of ids? could we cast to an array here so that we use the foreach anyway and same comment but about `$contactDetails` in L1673

mattwire replied:
Historically both $contactIds and $contactDetails had to be set which was a bit silly. This PR does away with that.  We are assuming that both have to be passed in as an array (and that's specified in the docblock) - I think that is sensible, rather than trying to accept non-array values.  Ideally we could drop one ($contactDetails) altogether but that means changing the function signature which will break stuff. @michaelmcandrew checking you agree.

michaelmcandrew replied:
hey there, no strong feelings. i guess seamus's comment stems from the fact that you are using is_array(). I am presuming you meant it in the sense of  'is array as opposed to null'. In other words, seems fine as is to me, but I coudl be missing some context.

mattwire commented:
Jenkins test this please

mattwire commented:
test this please

eileenmcnaughton commented:
@mattwire test failure error is PHP Fatal error:  require_once(): Failed opening required 'CiviTest/CiviTestSMSProvider.php'

mattwire commented:
> @mattwire test failure error is PHP Fatal error: require_once(): Failed opening required 'CiviTest/CiviTestSMSProvider.php'
Ok, rebased and forgot to include the file.  Hopefully tests will pass now...

mattwire commented:
@michaelmcandrew The tests are failing now because of the new send SMS permission I think. Any idea how to fix that?

eileenmcnaughton commented:
@mattwire @michaelmcandrew I think it would be normal to be able to call a function with $params['check_permissions'] = 0 - but unless it's accessible from the api that is optional. The other alternative is a test helper like the existing...

```
  /**
   * @expectedException CRM_Core_Exception
   * @expectedExceptionMessage You do not have the 'send SMS' permission
   */
  public function testSendSMSWithoutPermission() {
    $dummy = NULL;
    $session = CRM_Core_Session::singleton();
    $config = &CRM_Core_Config::singleton();
    $config->userPermissionClass->permissions = array('access CiviCRM');

    CRM_Activity_BAO_Activity::sendSMS(
      $dummy,
      $dummy,
      $dummy,
      $dummy,
      $session->get('userID')
    );
  }
```

michaelmcandrew commented:
Just an FYI, I am on holiday for the next week...

eileenmcnaughton commented:
@michaelmcandrew I didn't authorize that

seamuslee001 commented on tests/phpunit/CRM/Activity/BAO/ActivityTest.php:
```diff
@@ -1154,4 +1154,140 @@ public function testSendSMSWithoutPermission() {
+    // Get a "logged in" user to set as source of Sms.
+    $session = CRM_Core_Session::singleton();
+    $sourceContactId = $session->get('userID');
+
```
@mattwire you could here add 

```
     $config = &CRM_Core_Config::singleton();
     $config->userPermissionClass->permissions = array('access CiviCRM', 'send SMS');
```

eileenmcnaughton replied:
or, better 

CRM_Core_Config::singleton()->userPermissionClass->permissions = array('access CiviCRM', 'send SMS');

(that &CRM.... is C & P from somewhere)

mattwire commented:
test this please

mattwire commented:
test this please

eileenmcnaughton commented:
@seamuslee001 you reviewed this previously & were OK with it. Related test fails are now fixed - is it mergeable now? (unrelated test fail exists)

eileenmcnaughton commented:
test this please

seamuslee001 commented:
(*CiviCRM Review Template WORD-1.1*)

<!-- In each category, change the word "Undecided" to "Pass" or "Issue". Add explanatory comments if prompted or desired. -->

* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass Reviewed the code in depth and are happy with the changes and standardised coding practices
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass

seamuslee001 commented:
Merging
